src/wallet3/db_schema.cpp

```diff
@@ -420,5 +425,66 @@ namespace wallet
   {
     return prepared_get<int64_t>("SELECT output_count FROM metadata WHERE id=0;");
   }
+  void
+  WalletDB::save_keys(std::shared_ptr<Keyring> keys)
+  {
+    std::string query = "SELECT spend_priv, spend_pub, view_priv, view_pub FROM metadata WHERE id=0";
+    auto st = prepared_st(query);
+    std::string spend_priv_str;
+    std::string spend_pub_str;
+    std::string view_priv_str;
+    std::string view_pub_str;
+    while (st->executeStep())
+    {
+      auto from_db = db::get<std::string, std::string, std::string, std::string>(st);
+      spend_priv_str = std::get<0>(from_db);
+      spend_pub_str = std::get<1>(from_db);
+      view_priv_str = std::get<2>(from_db);
+      view_pub_str = std::get<3>(from_db);
+    }
+    if (spend_priv_str != "" || spend_pub_str != "" || view_priv_str != "" || view_pub_str != "")
+    {
+      crypto::secret_key spend_priv;
+      crypto::public_key spend_pub;
+      crypto::secret_key view_priv;
+      crypto::public_key view_pub;
+      tools::hex_to_type<crypto::secret_key>(spend_priv_str, spend_priv);
+      tools::hex_to_type<crypto::public_key>(spend_pub_str, spend_pub);
+      tools::hex_to_type<crypto::secret_key>(view_priv_str, view_priv);
+      tools::hex_to_type<crypto::public_key>(view_pub_str, view_pub);
+      auto keyring = Keyring(spend_priv, spend_pub, view_priv, view_pub, keys->nettype);
+      if (keyring.get_main_address() != keys->get_main_address())
+        throw std::runtime_error("loaded keys do not match database file");
+
+    } else {
+      auto account_keys = keys->export_keys();
+      prepared_exec("UPDATE metadata SET spend_priv = ?, spend_pub = ?, view_priv = ?, view_pub = ? where id = 0;",
+          tools::type_to_hex(account_keys.m_spend_secret_key),
```

Comment:
could move these out of the if/else and replace the `if` section above with

```
auto loaded_spend_priv_str = tools::type_to_hex(account_keys.m_spend_secret_key);
if ((not spend_priv_str.empty()) and spend_priv_str != loaded_spend_priv_str)
  throw ...;
```

for each of the keys.

That would also allow for, for example, the pubkeys to be stored in the db but not the privkeys.  `Keyring::export_keys` would have to change (or a new method like `export_public_keys` added) and this method would need an argument whether to store the priv keys but it's probably a good idea.

Reply:
Database should also not need to worry about the keyring class